feat: Add elasticsearch toolkit for conversational search

zach-shu opened the pull request:
This PR adds a new conversational search toolkit using Elasticsearch. 

Signed off by: zhongzheng.shu@ibm.com

jwm4 commented on integrations/extensions/docs/elasticsearch-install-and-setup/ICD_Elasticsearch_install_and_setup.md:
```diff
@@ -0,0 +1,250 @@
+
+
+## Step 2: Set up Kibana to connect to Elasticsearch
+* Install [Docker](https://www.docker.com/) so that you can pull the Kibana container image later.
```
This is saying to install Docker.  Is this a reference to Docker Desktop or something else (sorry, I know very little about container management software)?

My understanding is that this is not allowed for IBM employees as per https://w3.ibm.com/w3publisher/third-party-software/strategic-software-migration/software-development-tools .  Is it possible to do these steps using an alternative tool such as the ones listed on that page?  There are more details at https://w3.ibm.com/w3publisher/docker-desktop .

zach-shu replied:
Good point! I added a comment there to install `Colima` if we don't want to use `Docker Desktop`. All the other steps stay the same.

jwm4 commented on integrations/extensions/starter-kits/elasticsearch/README.md:
````diff
@@ -0,0 +1,40 @@
+    }
+    ```
+  * Try typing in anything in your preview chat to trigger `No action matches` action. 
+    If you see a successful extension call with valid response in the debug view, your Elasticsearch custom extension has been set up successfully.
````
Unlike all of our other search kits, this kit has no actions.json file showing how to use it to do basic search.  Ideally, we would have one that is comparable to the ones in the other kit.  However, basic search is not really a priority right now, so a fair compromise might be to just document the limitation, e.g., something like this:

> This kit does not provide a sample actions json file that shows how to use the kit in a real assistant.  There is a simple example that shows how to setup your No Action Matches action to call the search on a single fixed query, but no instructions for making it work with user queries.  The reason we have left this out is that most of the interest we've been seeing lately has not been from plain search but rather from search combined with generative AI to produce concise answers, which is sometimes called [retrieval-augmented generation](https://research.ibm.com/blog/retrieval-augmented-generation-RAG) or conversational search.  We have sample actions for doing this with Elasticsearch in our [conversational search starter kit](../language-model-conversational-search/README.md#example-1-connect-your-assistant-to-elasticsearch-and-watsonx-via-custom-extensions).  That kit builds on this one and adds a connection to watsonx.ai to generate answers from the search results.

jwm4 commented on integrations/extensions/starter-kits/elasticsearch/README.md:
```diff
@@ -0,0 +1,40 @@
+## Install and set up Elasticsearch
+
+Before starting, you will need to provision a Elasticsearch instance and set up a search index. Please follow the steps in 
+[elasticsearch-install-and-setup-guide](../../docs/elasticsearch-install-and-setup/ICD_Elasticsearch_install_and_setup.md) for more details.
```
OpenAPI spec should be in the /elasticsearch folder, not the /elasticsearch/assets folder for consistency with other search kits.

jwm4 commented on integrations/extensions/starter-kits/elasticsearch/README.md:
```diff
@@ -0,0 +1,40 @@
+## Build a custom extension in watsonx Assistant with Elasticsearch API
+
+* Download the OpenAPI specification [elasticsearch-generic-openapi.json](assets/elasticsearch-generic-openapi.json)
+* Use the OpenAPI specification to [build a custom extension](https://cloud.ibm.com/docs/watson-assistant?topic=watson-assistant-build-custom-extension#building-the-custom-extension), name it `Elasticsearch`
```
This says to name it `Elasticsearch` but the image below shows you named yours `ElasticSearch`.  Really users can name this whatever they want.  I recommend just deleting `, name it Elasticsearch` from this README and just not commenting on what to name it.

jwm4 commented on integrations/extensions/starter-kits/language-model-conversational-search/README.md:
```diff
@@ -17,7 +18,79 @@ All examples in this starter kit require that you use the [new IBM watsonx Assis
+- `query_text`: You MAY change this to pass queries to Watson Discovery. By default the Search action passes the user’s input.text directly.
+- `search_results`: Response object from Elasticsearch search query. It is set and used by the Search action.
+- `snippet` : Top results from the Watson Discovery document search.
+- `verbose`: A boolean that will print debugging output if true. Default is false. Note that if you turn this on, you will see the prompt we send to the model rendered as if it were HTML, which looks messy (all the text is crossed out because the start of the prompt is a tag that Llama uses to represent the start of a message but represents strike-through in HTML).
```
`verbose` is set to `true` in the Actions json file.  It should be set to `false` -- that's what's stated here in the README, and it is a much better experience for new users to only get all the verbose output when they explicitly ask for it.

jwm4 commented on integrations/extensions/starter-kits/language-model-conversational-search/README.md:
```diff
@@ -17,7 +18,79 @@ All examples in this starter kit require that you use the [new IBM watsonx Assis
+- `snippet` : Top results from the Watson Discovery document search.
+- `verbose`: A boolean that will print debugging output if true. Default is false. Note that if you turn this on, you will see the prompt we send to the model rendered as if it were HTML, which looks messy (all the text is crossed out because the start of the prompt is a tag that Llama uses to represent the start of a message but represents strike-through in HTML).
+- `watsonx_api_version` - watsonx api date version. It currently defaults to `2023-05-29`.
+- `watsonx_project_id`: You **MUST** set this value to be [a project ID value from watsonx](https://dataplatform.cloud.ibm.com/docs/content/wsj/manage-data/manage-projects.html). By default, this is a [sandbox project id](https://dataplatform.cloud.ibm.com/docs/content/wsj/manage-data/sandbox.html) that is automatically created for you when you sign up for watsonx.ai.
```
`watsonx_project_id` is set to some value in the Actions json file.  It should not be set in the json file -- the user needs to set it.

Also, the Discovery->watsonx kit has a very nice check to see if `watsonx_project_id` is set.  If it is not, it provides instructions for doing so.  I would recommend following that approach.

zach-shu replied:
I removed the default value for `watsonx_project_id` in the workspace JSON file. I couldn't find what you said about ```the Discovery->watsonx kit has a very nice check to see if watsonx_project_id is set```. But without the default value, any call to the watsonx extension would fail with a message saying that project_id should be set.

jwm4 commented on integrations/extensions/starter-kits/language-model-conversational-search/README.md:
```diff
@@ -17,7 +18,79 @@ All examples in this starter kit require that you use the [new IBM watsonx Assis
+- `model_parameters_temperature` : The value used to control the next token probabilities. The range is from 0 to 1.00; 0 makes it deterministic.
+- `model_response`: The text generated by the model in response to the model input.
+- `passages` : Concatenation of top search results.
+- `query_body`: The query body that will be sent to Elasticsearch search index. The query body is set and used by the Search action.
```
The query body in step 3 of the Search action assumes that you have an ELSER field in your schema named `text_expansion` and that you want to do an ELSER search on that field only.  If you want to do an ELSER search on some other field, a non-ELSER search, or a hybrid search, you need to update the query body assignment in step 3 of the Search action.  I would mention this in the documentation and maybe provide an assortment of reasonable example values.  It should probably go in another section, because it would get lost as just extra sentences in this list.

zach-shu replied:
Good point. I added more query body examples/references for different types of queries here https://github.com/watson-developer-cloud/assistant-toolkit/pull/233/files#diff-04d57dc0d43f09b15d9d739d5b22ae2799d75c45151c56605a020ad6be22e3deR29-R51

jwm4 commented on integrations/extensions/starter-kits/language-model-conversational-search/README.md:
```diff
@@ -17,7 +18,79 @@ All examples in this starter kit require that you use the [new IBM watsonx Assis
+| Action                        | Description                                                                                                                                                                                                                                                                                                  |
+| ----------------------------- |--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Search                        | Connects to Elasticsearch to search for documents related to the user query. The "No Action Matches" action has been configured to send all input to this action, so whatever the user enters will be used as the search input. It invokes the "Generate Answer" action to generate a response to the query. |
+| Generate Answer               | Configures the query prompt and document passages resulting from search, and calls the action "Invoke watsonx generation API". It is not meant to be invoked directly, but rather by the "Search" action.                                                                                                    |
```
The Generate Answer action uses the _full_ text of the search results and sends it along to the models.  This works well if the documents in your index are quite short.  That can be true if your original source documents are very short _or_ you have split your original source documents into smaller chunks and have indexed those chunks as separate documents in Elastic.  If not, you might need to provide a query body that asks for highlights from Elasticsearch and update the logic of Generate Answer to use the highlights and not the full text.  I don't think we want to actually _do_ that in this kit, but we should mention it in the README.

zach-shu replied:
I added some more description about the sample data we used in this kit [here](https://github.com/watson-developer-cloud/assistant-toolkit/pull/233/files#diff-40d59f42b193719a41b6d5644f417ea9362b5922595cb4144268b6610297e39bR132-R137).

I am not sure if mentioning the Elasticsearch `Highlighting` feature in this starter kit is a good idea. It could be confusing to the user if there are too many paths/options in the starter kit, IMO.

jwm4 reviewed (changes requested):
Various changes requested.  Note that I attached them to the relevant sections of the README for clarity but in some cases the requested change is really to the actions JSON file.

zach-shu commented:
Hi @jwm4, thanks a lot for your review. I have made some changes according to your comments. Could you please take a look again?

jwm4 commented on integrations/extensions/docs/elasticsearch-install-and-setup/ICD_Elasticsearch_install_and_setup.md:
```diff
@@ -0,0 +1,257 @@
+      <img src="assets/config_result_content.png" width="530" height="360" />  
+      Note: `author` and `name` are two fields from the documents in the index created from the previous step. They will be presented in the search result or used by Conversational Search.
+
+* Use the Search extension for `No action matches` scenario
```
These instructions work, but they are not quite the recommended approach, which involves adding a new step and then removing the existing first step.  The recommended approach is documented at https://cloud.ibm.com/docs/watson-assistant?topic=watson-assistant-search-add#search-no-action-matches .  Soon it will be replaced by new functionality that enables search integration _automatically_, but that might still be a few sprints away.  For now, I would drop this entire section and replace it with a link to https://cloud.ibm.com/docs/watson-assistant?topic=watson-assistant-search-add#search-no-action-matches

FWIW, the main reason the recommended approach is better is that the approach in this document keeps the "I'm afraid I don't understand. Please rephrase your question." message and puts it before each search result, as you can see in the screenshot.

jwm4 reviewed (changes requested):
Found only one minor issue with section 4 (but addressing it will require new screenshots, I think).  I will start reviewing section 5 next.

jwm4 commented on integrations/extensions/starter-kits/elasticsearch/README.md:
```diff
@@ -0,0 +1,73 @@
+## Install and set up Elasticsearch
+
+Before starting, you will need to provision a Elasticsearch instance and set up a search index. Please follow the steps in 
+[elasticsearch-install-and-setup-guide](../../docs/elasticsearch-install-and-setup/ICD_Elasticsearch_install_and_setup.md) for more details.
```
The link points to our guide for setting up Elastic on IBM Cloud.  We should update this sentence to something like this:

> If you are using Elasticsearch on IBM Cloud, you can follow the steps in our
[IBM Cloud Elasticsearch install and setup guide](../../docs/elasticsearch-install-and-setup/ICD_Elasticsearch_install_and_setup.md) for more details.  If you are using Elasticsearch on some other cloud or network, that should also work with this starter kit, but you will need to find set-up instructions appropriate to that environment.

zach-shu commented:
Hi @jwm4, thanks again for your review. I think I have addressed all the comments here. Do you mind taking another look at this PR?

As we discussed about adding an action workspace JSON for the Elasticsearch starter-kit, I agree and will add it in a follow-up PR.

jwm4 reviewed (approved):
This looks good now.  Thanks!
